docs/tech-notes/range-merges.md

```diff
@@ -0,0 +1,259 @@
+# Range merges
+
+**Last update:** December 21, 2018
+
+**Original author:** Nikhil Benesch
+
+This document serves as an end-to-end description of the implementation of range
+merges. The intended audience is someone who is reasonably familiar with core
+but unfamiliar with either the "how" or the "why" of the range merge
+implementation.
+
+The most complete documentation, of course, is in the code, tests, and the
+surrounding comments, but those pieces are necessarily split across several
+files and packages. That scattered knowledge is centralized here, without
+excessive detail that is likely to become stale.
+
+## Overview
+
+A range merge begins when two adjacent ranges are selected to be merged
+together. For example, suppose our adjacent ranges are _Q_ and _R_, somewhere in
+the middle of the keyspace:
+
+```
+--+-----+-----+--
+  |  Q  |  R  |
+--+-----+-----+--
+```
+
+We'll call _Q_ the left-hand side (LHS) of the merge, and _R_ the right-hand
+side (RHS) of the merge. For reasons that will become clear later, we also refer
+to _Q_ as the subsuming range and _R_ as the subsumed range.
+
+The merge is coordinated by the LHS. The coordinator begins by verifying that a)
+the two ranges are, in fact, adjacent, and b) that the replica sets of the two
+ranges are aligned. Replica set alignment is a term that is currently only
+relevant to merges; it means that the set of stores with replicas of the LHS
+exactly matches the set of stores with replicas of the RHS. For example, this
+replica set is aligned:
+
+```
+Store 1    Store 2     Store 3     Store 4
++-----+    +-----+     +-----+     +-----+
+| P Q |    | P Q |     |     |     | P Q |
++-----+    +-----+     +-----+     +-----+
+```
+
+By requiring replica set alignment, the merge operation is reduced to a
+metadata update, albiet a tricky one, as the stores that will have a copy of
+the merged range PQ already have all the constitudent data, by virtue of having
+a copy of both P and Q before the merge begins.
+
+After verifying that the merge is sensible, the coordinator transactionally
+updates the implicated range descriptors, adjusting P's range descriptor so that
+it extends to Q's end and deleting Q's range descriptor.
+
+Then, the coordinator needs to atomically move responsibility for the data in
+the RHS to the LHS. This is tricky, as the lease on the LHS may be held by a
+different store than the lease on the RHS, The coordinator notifies the RHS that
+it is about to be subsumed and it is prohibited from serving any additional read
+or write traffic. Only when the coordinator has received an acknowledgement from
+_every_ replica of the RHS, indicating that no traffic is possibly being served
+on the RHS, does the coordinator commit the merge transaction.
```

Comment:
Not a huge deal, but I take it that this means merges don't work 3-node clusters with one node down, similar to how the lack of quiescence can cause massive performance hits in such clusters.

Reply:
Yep, that's right. There's a theoretical that you could merge with a simple majority, but it didn't seem worth it. It would be massively complicated. Not sure if I'll have time to write it up formally, so let me know if you're curious and I can scrawl it down in a comment at least.